nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -65,44 +63,100 @@ impl DataStore {
+
+        // TODO: Need auth check here. Only fleet viewers can list children on
+        // IP_POOL_LIST, so if we check that, nobody can make instances. This
+        // used to check CreateChild on an individual IP pool, but now we're not
+        // looking up by name so the check is more complicated
+        //
+        // opctx
+        //     .authorize(authz::Action::ListChildren, &authz::IP_POOL_LIST)
+        //     .await?;
+
+        dsl::ip_pool
+            .filter(dsl::silo_id.eq(authz_silo_id).or(dsl::silo_id.is_null()))
+            .filter(
+                dsl::project_id.eq(project_id).or(dsl::project_id.is_null()),
+            )
+            .filter(dsl::is_default.eq(true))
+            .filter(dsl::time_deleted.is_null())
+            // this will sort by most specific first, i.e.,
+            //
+            //   (silo, project)
+            //   (silo, null)
+            //   (null, null)
+            //
+            // then by only taking the first result, we get the most specific one
+            .order((
+                dsl::project_id.asc().nulls_last(),
+                dsl::silo_id.asc().nulls_last(),
+            ))
+            .select(IpPool::as_select())
+            .first_async::<IpPool>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
     }
 
-    /// Looks up an IP pool by name.
+    /// Looks up an IP pool by name if it does not conflict with your current scope.
     pub(crate) async fn ip_pools_fetch_for(
         &self,
         opctx: &OpContext,
-        action: authz::Action,
         name: &Name,
-    ) -> LookupResult<(authz::IpPool, IpPool)> {
+        project_id: Option<Uuid>,
+    ) -> LookupResult<IpPool> {
+        let authz_silo_id = opctx.authn.silo_required()?.id();
+
         let (.., authz_pool, pool) = LookupPath::new(opctx, &self)
             .ip_pool_name(&name)
-            .fetch_for(action)
+            // any authenticated user can CreateChild on an IP pool. this is
+            // meant to represent allocating an IP
+            .fetch_for(authz::Action::CreateChild)
             .await?;
-        if pool.internal {
-            return Err(authz_pool.not_found());
+
+        // You can't look up a pool by name if it conflicts with your current
+        // scope, i.e., if it has a silo or project and those are different from
+        // your current silo or project
+
+        if let Some(pool_silo_id) = pool.silo_id {
+            if pool_silo_id != authz_silo_id {
+                return Err(authz_pool.not_found());
+            }
         }
 
-        Ok((authz_pool, pool))
+        match (pool.project_id, project_id) {
+            (Some(pool_project_id), Some(current_project_id)) => {
+                if pool_project_id != current_project_id {
+                    return Err(authz_pool.not_found());
+                }
+            }
+            _ => {}
+        }
```

Comment:
I don't understand what you mean, could you elaborate?

Reply:
Just imagining something tying into #3995 which I imagine would allow augmenting the `LookupPath::new(opctx, &self).ip_pool_name(&name)` call with `.project_id(project_id)`